// @remotion/studio-server/dist/preview-server/routes/log-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logUpdate = exports.formatPropChange = exports.strikeThrough = exports.bg = exports.fg = exports.normalizeQuotes = exports.warnAboutPrettierOnce = void 0;
const renderer_1 = require("@remotion/renderer");
let warnedAboutPrettier = false;
const warnAboutPrettierOnce = (logLevel) => {
    if (warnedAboutPrettier) {
        return;
    }
    warnedAboutPrettier = true;
    renderer_1.RenderInternals.Log.warn({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.yellow('Could not format with Prettier. File will need to be formatted manually.'));
};
exports.warnAboutPrettierOnce = warnAboutPrettierOnce;
const normalizeQuotes = (str) => {
    if (str.length >= 2 &&
        ((str.startsWith("'") && str.endsWith("'")) ||
            (str.startsWith('"') && str.endsWith('"')))) {
        return `'${str.slice(1, -1)}'`;
    }
    return str;
};
exports.normalizeQuotes = normalizeQuotes;
// 24-bit ANSI helpers
const fg = (r, g, b, str) => `\u001b[38;2;${r};${g};${b}m${str}\u001b[39m`;
exports.fg = fg;
const bg = (r, g, b, str) => `\u001b[48;2;${r};${g};${b}m${str}\u001b[49m`;
exports.bg = bg;
// eslint-disable-next-line no-control-regex
const stripAnsi = (str) => str.replace(/\u001b\[[0-9;]*m/g, '');
const strikeThrough = (str) => `\u001b[9m\u001b[38;2;255;85;85m${stripAnsi(str)}\u001b[39m\u001b[29m`;
exports.strikeThrough = strikeThrough;
// Monokai-inspired syntax colors
const attrName = (str) => (0, exports.fg)(166, 226, 46, str);
const equals = (str) => (0, exports.fg)(249, 38, 114, str);
const punctuation = (str) => (0, exports.fg)(248, 248, 242, str);
const stringValue = (str) => (0, exports.fg)(230, 219, 116, str);
const numberValue = (str) => (0, exports.fg)(174, 129, 255, str);
const colorValue = (str) => {
    if ((str.startsWith("'") && str.endsWith("'")) ||
        (str.startsWith('"') && str.endsWith('"'))) {
        return stringValue(str);
    }
    if (/^-?\d+(\.\d+)?$/.test(str)) {
        return numberValue(str);
    }
    return punctuation(str);
};
const colorEnabled = () => renderer_1.RenderInternals.chalk.enabled();
// Format key={value} with Monokai syntax highlighting
const formatSimpleProp = (key, value) => {
    return `${attrName(key)}${equals('=')}${punctuation('{')}${colorValue(value)}${punctuation('}')}`;
};
// Format parentKey={{childKey: value}} with Monokai syntax highlighting
const formatNestedProp = (parentKey, childKey, value) => {
    return `${attrName(parentKey)}${equals('=')}${punctuation('{{')}${punctuation(childKey)}${punctuation(':')} ${colorValue(value)}${punctuation('}}')}`;
};
const formatSideProps = ({ removedProps, addedProps, }) => {
    const parts = [];
    for (const { valueString } of removedProps) {
        parts.push(colorEnabled() ? (0, exports.strikeThrough)(valueString) : valueString);
    }
    for (const { valueString } of addedProps) {
        parts.push(valueString);
    }
    if (parts.length === 0) {
        return '';
    }
    return `, ${parts.join(', ')}`;
};
const formatPropChange = ({ key, oldValueString, newValueString, defaultValueString, removedProps, addedProps, }) => {
    const suffix = formatSideProps({ removedProps, addedProps });
    if (!colorEnabled()) {
        const dotIdx = key.indexOf('.');
        if (dotIdx === -1) {
            return `${key}={${oldValueString}} \u2192 ${key}={${newValueString}}${suffix}`;
        }
        const parent = key.slice(0, dotIdx);
        const child = key.slice(dotIdx + 1);
        return `${parent}={{${child}: ${oldValueString}}} \u2192 ${parent}={{${child}: ${newValueString}}}${suffix}`;
    }
    const dotIndex = key.indexOf('.');
    const formatProp = (value) => dotIndex === -1
        ? formatSimpleProp(key, value)
        : formatNestedProp(key.slice(0, dotIndex), key.slice(dotIndex + 1), value);
    if (defaultValueString !== null && newValueString === defaultValueString) {
        return `${(0, exports.strikeThrough)(formatProp(oldValueString))}${suffix}`;
    }
    if (defaultValueString !== null && oldValueString === defaultValueString) {
        return `${formatProp(newValueString)}${suffix}`;
    }
    return `${formatProp(oldValueString)} \u2192 ${formatProp(newValueString)}${suffix}`;
};
exports.formatPropChange = formatPropChange;
const logUpdate = ({ fileRelativeToRoot, line, key, oldValueString, newValueString, defaultValueString, formatted, logLevel, removedProps, addedProps, }) => {
    const locationLabel = `${fileRelativeToRoot}:${line}`;
    const propChange = (0, exports.formatPropChange)({
        key,
        oldValueString: (0, exports.normalizeQuotes)(oldValueString),
        newValueString: (0, exports.normalizeQuotes)(newValueString),
        defaultValueString: defaultValueString !== null ? (0, exports.normalizeQuotes)(defaultValueString) : null,
        removedProps,
        addedProps,
    });
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}:`)} ${propChange}`);
    if (!formatted) {
        (0, exports.warnAboutPrettierOnce)(logLevel);
    }
};
exports.logUpdate = logUpdate;
