// @remotion/studio-server/dist/preview-server/routes/save-sequence-props.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveSequencePropsHandler = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const renderer_1 = require("@remotion/renderer");
const remotion_1 = require("remotion");
const get_all_schema_keys_1 = require("../../codemods/get-all-schema-keys");
const update_sequence_props_1 = require("../../codemods/update-sequence-props/update-sequence-props");
const file_watcher_1 = require("../../file-watcher");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const format_prop_change_1 = require("./log-updates/format-prop-change");
const log_update_1 = require("./log-updates/log-update");
const saveSequencePropsHandler = async ({ input: { fileName, nodePath, key, value, defaultValue, schema }, remotionRoot, logLevel, }) => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[save-sequence-props] Received request for fileName="${fileName}" key="${key}"`);
    const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
    const fileRelativeToRoot = node_path_1.default.relative(remotionRoot, absolutePath);
    if (fileRelativeToRoot.startsWith('..')) {
        throw new Error('Cannot modify a file outside the project');
    }
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const { output, oldValueStrings, formatted, logLine, removedProps } = await (0, update_sequence_props_1.updateSequenceProps)({
        input: fileContents,
        nodePath,
        updates: [
            {
                key,
                value: JSON.parse(value),
                defaultValue: defaultValue !== null ? JSON.parse(defaultValue) : null,
            },
        ],
        schema: remotion_1.Internals.sequenceSchema,
    });
    const oldValueString = oldValueStrings[0];
    const newValueString = JSON.stringify(JSON.parse(value));
    const parsedDefault = defaultValue !== null ? JSON.parse(defaultValue) : null;
    const defaultValueString = parsedDefault !== null ? JSON.stringify(parsedDefault) : null;
    const normalizedOld = (0, log_update_1.normalizeQuotes)(oldValueString);
    const normalizedNew = (0, log_update_1.normalizeQuotes)(newValueString);
    const normalizedDefault = defaultValueString !== null ? (0, log_update_1.normalizeQuotes)(defaultValueString) : null;
    const undoPropChange = (0, format_prop_change_1.formatPropChange)({
        key,
        oldValueString: normalizedNew,
        newValueString: normalizedOld,
        defaultValueString: normalizedDefault,
        removedProps: [],
        addedProps: removedProps,
    });
    const redoPropChange = (0, format_prop_change_1.formatPropChange)({
        key,
        oldValueString: normalizedOld,
        newValueString: normalizedNew,
        defaultValueString: normalizedDefault,
        removedProps,
        addedProps: [],
    });
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'sequence-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output);
    (0, log_update_1.logUpdate)({
        fileRelativeToRoot,
        line: logLine,
        key,
        oldValueString,
        newValueString,
        defaultValueString,
        formatted,
        logLevel,
        removedProps,
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const newStatus = (0, can_update_sequence_props_1.computeSequencePropsStatus)({
        fileName,
        keys: (0, get_all_schema_keys_1.getAllSchemaKeys)(schema),
        nodePath,
        remotionRoot,
    });
    return newStatus;
};
exports.saveSequencePropsHandler = saveSequencePropsHandler;
