// @remotion/transitions/dist/esm/wipe.mjs
// src/presentations/wipe.tsx
import { useMemo } from "react";
import { AbsoluteFill } from "remotion";
import { jsx } from "react/jsx-runtime";
var makePolygonIn = (progress, direction) => {
  const p = progress * 100;
  switch (direction) {
    case "from-left":
      return `polygon(0% 0%, ${p}% 0%, ${p}% 100%, 0% 100%)`;
    case "from-top-left":
      return `polygon(0% 0%, ${p * 2}% 0%, 0% ${p * 2}%)`;
    case "from-top":
      return `polygon(0% 0%, 100% 0%, 100% ${p}%, 0% ${p}%)`;
    case "from-top-right":
      return `polygon(100% 0%, ${100 - p * 2}% 0%, 100% ${p * 2}%)`;
    case "from-right":
      return `polygon(100% 0%, 100% 100%, ${100 - p}% 100%, ${100 - p}% 0%)`;
    case "from-bottom-right":
      return `polygon(100% 100%, ${100 - p * 2}% 100%, 100% ${100 - p * 2}%)`;
    case "from-bottom":
      return `polygon(0% 100%, 100% 100%, 100% ${100 - p}%, 0% ${100 - p}%)`;
    case "from-bottom-left":
      return `polygon(0% 100%, 0% ${100 - p * 2}%, ${p * 2}% 100%)`;
    default:
      throw new Error(`Unknown direction ${JSON.stringify(direction)}`);
  }
};
var makePolygonOut = (progress, direction) => {
  const p = progress * 100;
  switch (direction) {
    case "from-left":
      return `polygon(100% 100%, ${100 - p}% 100%, ${100 - p}% 0%, 100% 0%)`;
    case "from-top-left":
      return `polygon(100% 100%, ${100 - p * 2}% 100%, 100% ${100 - p * 2}%)`;
    case "from-top":
      return `polygon(100% 100%, 0% 100%, 0% ${100 - p}%, 100% ${100 - p}%)`;
    case "from-top-right":
      return `polygon(0% 100%, ${p * 2}% 100%, 0% ${100 - p * 2}%)`;
    case "from-right":
      return `polygon(0% 0%, ${p}% 0%, ${p}% 100%, 0% 100%)`;
    case "from-bottom-right":
      return `polygon(0% 0%, ${p * 2}% 0%, 0% ${p * 2}%)`;
    case "from-bottom":
      return `polygon(100% 0%, 0% 0%, 0% ${p}%, 100% ${p}%)`;
    case "from-bottom-left":
      return `polygon(100% 0%, ${100 - p * 2}% 0%, 100% ${p * 2}%)`;
    default:
      throw new Error(`Unknown direction ${JSON.stringify(direction)}`);
  }
};
var WipePresentation = ({
  children,
  presentationProgress,
  presentationDirection,
  passedProps: {
    direction = "from-left",
    innerEnterStyle,
    innerExitStyle,
    outerEnterStyle,
    outerExitStyle
  }
}) => {
  const progressInDirection = presentationDirection === "entering" ? presentationProgress : 1 - presentationProgress;
  const clipPath = presentationDirection === "entering" ? makePolygonIn(progressInDirection, direction) : makePolygonOut(progressInDirection, direction);
  const style = useMemo(() => {
    return {
      width: "100%",
      height: "100%",
      justifyContent: "center",
      alignItems: "center",
      clipPath,
      ...presentationDirection === "entering" ? innerEnterStyle : innerExitStyle
    };
  }, [clipPath, innerEnterStyle, innerExitStyle, presentationDirection]);
  const outerStyle = useMemo(() => {
    return presentationDirection === "entering" ? outerEnterStyle : outerExitStyle;
  }, [outerEnterStyle, outerExitStyle, presentationDirection]);
  return /* @__PURE__ */ jsx(AbsoluteFill, {
    style: outerStyle,
    children: /* @__PURE__ */ jsx(AbsoluteFill, {
      style,
      children
    })
  });
};
var wipe = (props) => {
  return {
    component: WipePresentation,
    props: props ?? {}
  };
};
export {
  wipe
};
