// @remotion/transitions/dist/esm/zoom-blur.mjs
// src/html-in-canvas-presentation.tsx
import { useLayoutEffect, useMemo, useRef, useState, useCallback } from "react";
import { HtmlInCanvas, useDelayRender } from "remotion";
import { AbsoluteFill, Internals, useCurrentFrame } from "remotion";
import { jsx } from "react/jsx-runtime";
var HtmlInCanvasPresentation = ({
  children,
  onElementImage,
  onUnmount,
  presentationProgress,
  presentationDirection,
  shader,
  _experimentalEffects,
  passedProps,
  bothEnteringAndExiting
}) => {
  if (!HtmlInCanvas.isSupported()) {
    throw new Error("HTML in Canvas is not supported. Open this page in Chrome Canary with chrome://flags/#canvas-draw-element enabled.");
  }
  const canvasRef = useRef(null);
  const canvasSubtreeStyle = useMemo(() => {
    return {
      width: "100%",
      height: "100%",
      position: "absolute",
      top: 0,
      left: 0,
      right: 0,
      bottom: 0
    };
  }, []);
  const [offscreenCanvas] = useState(() => new OffscreenCanvas(1, 1));
  const passedPropsRef = useRef(passedProps);
  passedPropsRef.current = passedProps;
  const frame = useCurrentFrame();
  const frameRef = useRef(frame);
  frameRef.current = frame;
  const memoizedEffects = Internals.useMemoizedEffects(Internals.flattenEffects(_experimentalEffects ?? []));
  const effectsRef = useRef(memoizedEffects);
  effectsRef.current = memoizedEffects;
  const [instance] = useState(() => shader(offscreenCanvas));
  useLayoutEffect(() => {
    return () => {
      instance.cleanup();
    };
  }, [offscreenCanvas, instance]);
  const chainState = Internals.useEffectChainState();
  const { delayRender, continueRender } = useDelayRender();
  const draw = useCallback(async (prevImage, nextImage, progress) => {
    if (!canvasRef.current) {
      throw new Error("Canvas not found");
    }
    const handle = delayRender("onPaint");
    if (!prevImage && !nextImage) {
      continueRender(handle);
      instance.clear();
      return;
    }
    const width = prevImage?.width ?? nextImage?.width ?? 0;
    const height = prevImage?.height ?? nextImage?.height ?? 0;
    if (width === 0 || height === 0) {
      continueRender(handle);
      instance.clear();
      return;
    }
    offscreenCanvas.width = width;
    offscreenCanvas.height = height;
    instance.draw({
      prevImage,
      nextImage,
      width,
      height,
      time: progress,
      passedProps: passedPropsRef.current
    });
    await Internals.runEffectChain({
      state: chainState.get(width, height),
      source: offscreenCanvas,
      effects: effectsRef.current ?? [],
      frame: frameRef.current,
      width,
      height,
      output: canvasRef.current
    });
    continueRender(handle);
  }, [chainState, instance, offscreenCanvas, continueRender, delayRender]);
  const passThrough = bothEnteringAndExiting && presentationDirection === "exiting";
  useLayoutEffect(() => {
    if (passThrough) {
      return;
    }
    const canvas = canvasRef.current;
    if (!canvas) {
      throw new Error("Canvas not found");
    }
    canvas.layoutSubtree = true;
    const onPaint = () => {
      const firstChild = canvas.firstChild;
      if (!firstChild) {
        return;
      }
      const elementImage = canvas.captureElementImage(firstChild);
      onElementImage(elementImage, draw);
    };
    canvas.addEventListener("paint", onPaint);
    return () => {
      canvas.removeEventListener("paint", onPaint);
    };
  }, [onElementImage, presentationDirection, draw, passThrough]);
  useLayoutEffect(() => {
    if (passThrough) {
      return;
    }
    const canvas = canvasRef.current;
    if (!canvas) {
      throw new Error("Canvas not found");
    }
    canvas.requestPaint?.();
  }, [presentationProgress, passThrough, memoizedEffects]);
  useLayoutEffect(() => {
    if (passThrough) {
      return;
    }
    return () => {
      onUnmount();
    };
  }, [onUnmount, passThrough]);
  useLayoutEffect(() => {
    if (passThrough) {
      return;
    }
    const canvas = canvasRef.current;
    if (!canvas) {
      return;
    }
    const observer = new ResizeObserver(([entry]) => {
      canvas.width = entry.devicePixelContentBoxSize[0].inlineSize;
      canvas.height = entry.devicePixelContentBoxSize[0].blockSize;
    });
    observer.observe(canvas, { box: "device-pixel-content-box" });
  }, [passThrough]);
  if (passThrough) {
    return children;
  }
  return /* @__PURE__ */ jsx(AbsoluteFill, {
    children: /* @__PURE__ */ jsx("canvas", {
      ref: canvasRef,
      style: canvasSubtreeStyle,
      children
    })
  });
};
var makeHtmlInCanvasPresentation = (shader) => {
  const CompWithShader = (props) => {
    const { passedProps, ...otherProps } = props;
    const { _experimentalEffects, ...restPassedProps } = props.passedProps;
    return /* @__PURE__ */ jsx(HtmlInCanvasPresentation, {
      shader,
      passedProps: restPassedProps,
      _experimentalEffects,
      ...otherProps
    });
  };
  return (props) => {
    return {
      component: CompWithShader,
      props
    };
  };
};

// src/presentations/zoom-blur.tsx
var VERTEX_SHADER = `#version 300 es
in vec2 a_pos;
out vec2 v_uv;
void main() {
	v_uv = vec2(a_pos.x * 0.5 + 0.5, 0.5 - a_pos.y * 0.5);
	gl_Position = vec4(a_pos, 0.0, 1.0);
}`;
var FRAGMENT_SHADER = `#version 300 es
precision highp float;

uniform sampler2D u_prev;
uniform sampler2D u_next;
uniform float u_time;
uniform float u_aspect;
uniform float u_max_angle;

in vec2 v_uv;
out vec4 outColor;

const int SAMPLES = 16;
const float STRENGTH = 0.35;

vec2 transformUV(vec2 uv, float angle, float scale) {
	vec2 p = uv - 0.5;
	p.x *= u_aspect;
	p /= scale;
	float c = cos(-angle);
	float s = sin(-angle);
	p = vec2(p.x * c - p.y * s, p.x * s + p.y * c);
	p.x /= u_aspect;
	return p + 0.5;
}

float coverScale(float angle) {
	float c = abs(cos(angle));
	float s = abs(sin(angle));
	float ar = max(u_aspect, 1.0 / u_aspect);
	return c + ar * s;
}

vec4 zoomBlur(sampler2D tex, vec2 uv, float strength) {
	vec2 dir = uv - 0.5;
	vec4 acc = vec4(0.0);
	for (int i = 0; i < SAMPLES; i++) {
		float t = float(i) / float(SAMPLES - 1);
		float scale = 1.0 - strength * t;
		acc += texture(tex, 0.5 + dir * scale);
	}
	return acc / float(SAMPLES);
}

void main() {
	float mixT = u_time;

	float nextAngle = u_max_angle * mixT;
	float prevAngle = -u_max_angle * (1.0 - mixT);

	vec2 prevUV = transformUV(v_uv, prevAngle, coverScale(prevAngle));
	vec2 nextUV = transformUV(v_uv, nextAngle, coverScale(nextAngle));

	vec4 prev = zoomBlur(u_prev, prevUV, STRENGTH * (1.0 - mixT));
	vec4 next = zoomBlur(u_next, nextUV, STRENGTH * mixT);
	outColor = mix(prev, next, (1.0 - mixT));
}`;
var compileShader = (gl, source, type) => {
  const shader = gl.createShader(type);
  if (!shader) {
    throw new Error("Failed to create shader");
  }
  gl.shaderSource(shader, source);
  gl.compileShader(shader);
  if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
    const log = gl.getShaderInfoLog(shader);
    gl.deleteShader(shader);
    throw new Error(`Failed to compile shader: ${log}`);
  }
  return shader;
};
var createProgram = (gl) => {
  const program = gl.createProgram();
  if (!program) {
    throw new Error("Failed to create WebGL program");
  }
  const vs = compileShader(gl, VERTEX_SHADER, gl.VERTEX_SHADER);
  const fs = compileShader(gl, FRAGMENT_SHADER, gl.FRAGMENT_SHADER);
  gl.attachShader(program, vs);
  gl.attachShader(program, fs);
  gl.linkProgram(program);
  if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
    const log = gl.getProgramInfoLog(program);
    gl.deleteProgram(program);
    throw new Error(`Failed to link program: ${log}`);
  }
  return program;
};
var createTexture = (gl) => {
  const tex = gl.createTexture();
  if (!tex) {
    throw new Error("Failed to create texture");
  }
  gl.bindTexture(gl.TEXTURE_2D, tex);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
  gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, 1, 1, 0, gl.RGBA, gl.UNSIGNED_BYTE, new Uint8Array([0, 0, 0, 0]));
  return tex;
};
var zoomBlurShader = (canvas) => {
  const gl = canvas.getContext("webgl2", { premultipliedAlpha: true });
  if (!gl) {
    throw new Error("Failed to create WebGL2 context");
  }
  const program = createProgram(gl);
  const prevTex = createTexture(gl);
  const nextTex = createTexture(gl);
  const vao = gl.createVertexArray();
  gl.bindVertexArray(vao);
  const buffer = gl.createBuffer();
  gl.bindBuffer(gl.ARRAY_BUFFER, buffer);
  gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([-1, -1, 1, -1, -1, 1, 1, 1]), gl.STATIC_DRAW);
  const aPos = gl.getAttribLocation(program, "a_pos");
  gl.enableVertexAttribArray(aPos);
  gl.vertexAttribPointer(aPos, 2, gl.FLOAT, false, 0, 0);
  const uTime = gl.getUniformLocation(program, "u_time");
  const uPrev = gl.getUniformLocation(program, "u_prev");
  const uNext = gl.getUniformLocation(program, "u_next");
  const uAspect = gl.getUniformLocation(program, "u_aspect");
  const uMaxAngle = gl.getUniformLocation(program, "u_max_angle");
  const cleanup = () => {
    gl.deleteProgram(program);
    gl.deleteTexture(prevTex);
    gl.deleteTexture(nextTex);
  };
  const clear = () => {
    gl.clearColor(0, 0, 0, 0);
    gl.clear(gl.COLOR_BUFFER_BIT);
  };
  const draw = ({
    prevImage,
    nextImage,
    width,
    height,
    time,
    passedProps
  }) => {
    const { rotation = Math.PI / 6 } = passedProps;
    if (!prevImage && !nextImage) {
      return;
    }
    if (prevImage && (prevImage.width === 0 || prevImage.height === 0)) {
      return;
    }
    if (nextImage && (nextImage.width === 0 || nextImage.height === 0)) {
      return;
    }
    const effectiveTime = !prevImage ? 0 : !nextImage ? 1 : time;
    gl.viewport(0, 0, width, height);
    gl.clearColor(0, 0, 0, 0);
    gl.clear(gl.COLOR_BUFFER_BIT);
    gl.useProgram(program);
    gl.activeTexture(gl.TEXTURE0);
    gl.bindTexture(gl.TEXTURE_2D, prevTex);
    if (prevImage) {
      gl.texElementImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, prevImage);
    }
    gl.uniform1i(uPrev, 0);
    gl.activeTexture(gl.TEXTURE1);
    gl.bindTexture(gl.TEXTURE_2D, nextTex);
    if (nextImage) {
      gl.texElementImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, nextImage);
    }
    gl.uniform1i(uNext, 1);
    gl.uniform1f(uTime, effectiveTime);
    gl.uniform1f(uAspect, width / height);
    gl.uniform1f(uMaxAngle, rotation);
    gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
  };
  return {
    clear,
    cleanup,
    draw
  };
};
var zoomBlur = makeHtmlInCanvasPresentation(zoomBlurShader);
export {
  zoomBlurShader,
  zoomBlur
};
