// @remotion/transitions/dist/presentations/wipe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wipe = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const makePolygonIn = (progress, direction) => {
    const p = progress * 100;
    switch (direction) {
        case 'from-left':
            return `polygon(0% 0%, ${p}% 0%, ${p}% 100%, 0% 100%)`;
        case 'from-top-left':
            return `polygon(0% 0%, ${p * 2}% 0%, 0% ${p * 2}%)`;
        case 'from-top':
            return `polygon(0% 0%, 100% 0%, 100% ${p}%, 0% ${p}%)`;
        case 'from-top-right':
            return `polygon(100% 0%, ${100 - p * 2}% 0%, 100% ${p * 2}%)`;
        case 'from-right':
            return `polygon(100% 0%, 100% 100%, ${100 - p}% 100%, ${100 - p}% 0%)`;
        case 'from-bottom-right':
            return `polygon(100% 100%, ${100 - p * 2}% 100%, 100% ${100 - p * 2}%)`;
        case 'from-bottom':
            return `polygon(0% 100%, 100% 100%, 100% ${100 - p}%, 0% ${100 - p}%)`;
        case 'from-bottom-left':
            return `polygon(0% 100%, 0% ${100 - p * 2}%, ${p * 2}% 100%)`;
        default:
            throw new Error(`Unknown direction ${JSON.stringify(direction)}`);
    }
};
const makePolygonOut = (progress, direction) => {
    const p = progress * 100;
    switch (direction) {
        case 'from-left':
            return `polygon(100% 100%, ${100 - p}% 100%, ${100 - p}% 0%, 100% 0%)`;
        case 'from-top-left':
            return `polygon(100% 100%, ${100 - p * 2}% 100%, 100% ${100 - p * 2}%)`;
        case 'from-top':
            return `polygon(100% 100%, 0% 100%, 0% ${100 - p}%, 100% ${100 - p}%)`;
        case 'from-top-right':
            return `polygon(0% 100%, ${p * 2}% 100%, 0% ${100 - p * 2}%)`;
        case 'from-right':
            return `polygon(0% 0%, ${p}% 0%, ${p}% 100%, 0% 100%)`;
        case 'from-bottom-right':
            return `polygon(0% 0%, ${p * 2}% 0%, 0% ${p * 2}%)`;
        case 'from-bottom':
            return `polygon(100% 0%, 0% 0%, 0% ${p}%, 100% ${p}%)`;
        case 'from-bottom-left':
            return `polygon(100% 0%, ${100 - p * 2}% 0%, 100% ${p * 2}%)`;
        default:
            throw new Error(`Unknown direction ${JSON.stringify(direction)}`);
    }
};
const WipePresentation = ({ children, presentationProgress, presentationDirection, passedProps: { direction = 'from-left', innerEnterStyle, innerExitStyle, outerEnterStyle, outerExitStyle, }, }) => {
    const progressInDirection = presentationDirection === 'entering'
        ? presentationProgress
        : 1 - presentationProgress;
    const clipPath = presentationDirection === 'entering'
        ? makePolygonIn(progressInDirection, direction)
        : makePolygonOut(progressInDirection, direction);
    const style = (0, react_1.useMemo)(() => {
        return {
            width: '100%',
            height: '100%',
            justifyContent: 'center',
            alignItems: 'center',
            clipPath,
            ...(presentationDirection === 'entering'
                ? innerEnterStyle
                : innerExitStyle),
        };
    }, [clipPath, innerEnterStyle, innerExitStyle, presentationDirection]);
    const outerStyle = (0, react_1.useMemo)(() => {
        return presentationDirection === 'entering'
            ? outerEnterStyle
            : outerExitStyle;
    }, [outerEnterStyle, outerExitStyle, presentationDirection]);
    return (jsx_runtime_1.jsx(remotion_1.AbsoluteFill, { style: outerStyle, children: jsx_runtime_1.jsx(remotion_1.AbsoluteFill, { style: style, children: children }) }));
};
/*
 * @description A presentation where the entering slide slides over the exiting slide.
 * @see [Documentation](https://www.remotion.dev/docs/transitions/presentations/wipe)
 */
const wipe = (props) => {
    return {
        component: WipePresentation,
        props: props !== null && props !== void 0 ? props : {},
    };
};
exports.wipe = wipe;
