// @remotion/studio-shared/dist/schema-field-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFieldsToShow = exports.UNSUPPORTED_FIELD_ROW_HEIGHT = exports.SCHEMA_FIELD_ROW_HEIGHT = void 0;
const remotion_1 = require("remotion");
const no_react_1 = require("remotion/no-react");
exports.SCHEMA_FIELD_ROW_HEIGHT = 22;
exports.UNSUPPORTED_FIELD_ROW_HEIGHT = 22;
const SUPPORTED_SCHEMA_TYPES = new Set([
    'number',
    'boolean',
    'rotation',
    'translate',
    'enum',
]);
const getFieldsToShow = ({ getDragOverrides, getCodeValues, nodePath, schema, currentRuntimeValueDotNotation, }) => {
    const { merged: valuesDotNotation } = remotion_1.Internals.computeEffectiveSchemaValuesDotNotation({
        schema,
        currentValue: currentRuntimeValueDotNotation,
        overrideValues: getDragOverrides(nodePath),
        propStatus: getCodeValues(nodePath),
    });
    const activeSchema = remotion_1.Internals.flattenActiveSchema(schema, (key) => valuesDotNotation[key]);
    return Object.entries(activeSchema)
        .map(([key, fieldSchema]) => {
        const typeName = fieldSchema.type;
        const supported = SUPPORTED_SCHEMA_TYPES.has(typeName);
        if (typeName === 'hidden') {
            return null;
        }
        return {
            key,
            description: fieldSchema.description,
            typeName,
            supported,
            rowHeight: supported
                ? exports.SCHEMA_FIELD_ROW_HEIGHT
                : exports.UNSUPPORTED_FIELD_ROW_HEIGHT,
            currentRuntimeValue: currentRuntimeValueDotNotation[key],
            fieldSchema,
        };
    })
        .filter(no_react_1.NoReactInternals.truthy);
};
exports.getFieldsToShow = getFieldsToShow;
