// @remotion/studio/dist/helpers/calculate-timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateTimeline = void 0;
const get_sequence_visible_range_1 = require("./get-sequence-visible-range");
const get_timeline_nestedness_1 = require("./get-timeline-nestedness");
const get_timeline_sequence_hash_1 = require("./get-timeline-sequence-hash");
const get_timeline_sequence_sort_key_1 = require("./get-timeline-sequence-sort-key");
const sort_by_nonce_history_1 = require("./sort-by-nonce-history");
const getInheritedLoopDisplay = (sequence, sequences) => {
    if (sequence.loopDisplay) {
        return sequence.loopDisplay;
    }
    if (!sequence.parent) {
        return undefined;
    }
    const parent = sequences.find((s) => s.id === sequence.parent);
    if (!parent) {
        return undefined;
    }
    return getInheritedLoopDisplay(parent, sequences);
};
const calculateTimeline = ({ sequences, overrideIdsToNodePaths, }) => {
    var _a;
    var _b;
    const sortedSequences = (0, sort_by_nonce_history_1.sortItemsByNonceHistory)(sequences);
    const tracks = [];
    if (sortedSequences.length === 0) {
        return [];
    }
    const sameHashes = {};
    const hashesUsedInRoot = {};
    const cache = {};
    for (let i = 0; i < sortedSequences.length; i++) {
        const sequence = sortedSequences[i];
        if (!hashesUsedInRoot[sequence.rootId]) {
            hashesUsedInRoot[sequence.rootId] = [];
        }
        const actualHash = (0, get_timeline_sequence_hash_1.getTimelineSequenceHash)(sequence, sortedSequences, hashesUsedInRoot, cache);
        if (!sameHashes[actualHash]) {
            sameHashes[actualHash] = [];
        }
        sameHashes[actualHash].push(sequence.id);
        const cascadedStart = (0, get_sequence_visible_range_1.getCascadedStart)(sequence, sortedSequences);
        const visibleStart = (0, get_sequence_visible_range_1.getTimelineVisibleStart)(sequence, sortedSequences);
        const visibleDuration = (0, get_sequence_visible_range_1.getTimelineVisibleDuration)(sequence, sortedSequences);
        const overrideId = (_b = (_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.overrideId) !== null && _b !== void 0 ? _b : null;
        const nodePath = overrideId ? overrideIdsToNodePaths[overrideId] : null;
        tracks.push({
            sequence: {
                ...sequence,
                from: visibleStart,
                duration: visibleDuration,
                loopDisplay: sequence.type === 'audio' || sequence.type === 'video'
                    ? getInheritedLoopDisplay(sequence, sortedSequences)
                    : sequence.loopDisplay,
            },
            depth: (0, get_timeline_nestedness_1.getTimelineNestedLevel)(sequence, sortedSequences, 0),
            hash: actualHash,
            cascadedStart,
            cascadedDuration: sequence.duration,
            nodePathInfo: nodePath
                ? { nodePath, index: 0, numberOfSequencesWithThisNodePath: 0 }
                : null,
        });
    }
    const uniqueTracks = [];
    for (const track of tracks) {
        if (!uniqueTracks.find((t) => t.hash === track.hash)) {
            const { cascadedDuration, cascadedStart, ...cleanTrack } = track;
            uniqueTracks.push(cleanTrack);
        }
    }
    const nonceRanks = new Map();
    for (let i = 0; i < tracks.length; i++) {
        nonceRanks.set(tracks[i].sequence.id, i);
    }
    const sortedTracks = uniqueTracks.sort((a, b) => {
        const sortKeyA = (0, get_timeline_sequence_sort_key_1.getTimelineSequenceSequenceSortKey)(a, tracks, sameHashes, nonceRanks);
        const sortKeyB = (0, get_timeline_sequence_sort_key_1.getTimelineSequenceSequenceSortKey)(b, tracks, sameHashes, nonceRanks);
        return sortKeyA.localeCompare(sortKeyB);
    });
    const nodePathIndexCounters = new Map();
    return sortedTracks
        .map((track) => {
        var _a;
        if (track.nodePathInfo === null) {
            return track;
        }
        const key = track.nodePathInfo.nodePath.join('.');
        const index = (_a = nodePathIndexCounters.get(key)) !== null && _a !== void 0 ? _a : 0;
        nodePathIndexCounters.set(key, index + 1);
        return {
            ...track,
            nodePathInfo: {
                nodePath: track.nodePathInfo.nodePath,
                index,
                numberOfSequencesWithThisNodePath: 0,
            },
        };
    })
        .map((track) => {
        var _a;
        if (track.nodePathInfo === null) {
            return track;
        }
        const key = track.nodePathInfo.nodePath.join('.');
        return {
            ...track,
            nodePathInfo: {
                nodePath: track.nodePathInfo.nodePath,
                index: track.nodePathInfo.index,
                numberOfSequencesWithThisNodePath: (_a = nodePathIndexCounters.get(key)) !== null && _a !== void 0 ? _a : 0,
            },
        };
    });
};
exports.calculateTimeline = calculateTimeline;
