// @remotion/studio/dist/helpers/timeline-layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimelineLayerHeight = exports.TIMELINE_LAYER_HEIGHT_AUDIO = exports.TIMELINE_LAYER_HEIGHT_IMAGE = exports.TIMELINE_LAYER_HEIGHT_VIDEO = exports.getExpandedTrackHeight = exports.getTreeRowHeight = exports.flattenVisibleTreeNodes = exports.buildTimelineTree = exports.getEffectSchemaLabels = exports.EXPANDED_SECTION_PADDING_RIGHT = exports.EXPANDED_SECTION_PADDING_LEFT = exports.TREE_GROUP_ROW_HEIGHT = exports.TIMELINE_TRACK_EXPANDED_HEIGHT = exports.TIMELINE_ITEM_BORDER_BOTTOM = exports.TIMELINE_BORDER = exports.TIMELINE_PADDING = exports.getFieldsToShow = exports.UNSUPPORTED_FIELD_ROW_HEIGHT = exports.SCHEMA_FIELD_ROW_HEIGHT = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const no_react_1 = require("remotion/no-react");
const studio_shared_2 = require("@remotion/studio-shared");
Object.defineProperty(exports, "SCHEMA_FIELD_ROW_HEIGHT", { enumerable: true, get: function () { return studio_shared_2.SCHEMA_FIELD_ROW_HEIGHT; } });
Object.defineProperty(exports, "UNSUPPORTED_FIELD_ROW_HEIGHT", { enumerable: true, get: function () { return studio_shared_2.UNSUPPORTED_FIELD_ROW_HEIGHT; } });
Object.defineProperty(exports, "getFieldsToShow", { enumerable: true, get: function () { return studio_shared_2.getFieldsToShow; } });
exports.TIMELINE_PADDING = 16;
exports.TIMELINE_BORDER = 1;
exports.TIMELINE_ITEM_BORDER_BOTTOM = 1;
exports.TIMELINE_TRACK_EXPANDED_HEIGHT = 100;
exports.TREE_GROUP_ROW_HEIGHT = 22;
exports.EXPANDED_SECTION_PADDING_LEFT = 28;
exports.EXPANDED_SECTION_PADDING_RIGHT = 10;
const getEffectSchemaLabels = (effect) => {
    if (!effect.definition.schema) {
        return [];
    }
    return Object.entries(effect.definition.schema)
        .map(([key, fieldSchema]) => {
        if (fieldSchema.type === 'hidden') {
            return null;
        }
        return {
            key,
            description: fieldSchema.description,
        };
    })
        .filter(no_react_1.NoReactInternals.truthy);
};
exports.getEffectSchemaLabels = getEffectSchemaLabels;
const buildTimelineTree = ({ sequence, nodePathInfo, getDragOverrides, getCodeValues, }) => {
    var _a;
    const roots = [];
    const { nodePath, index } = nodePathInfo;
    if (sequence.effects.length > 0) {
        roots.push({
            kind: 'group',
            nodePathInfo: {
                nodePath: [...nodePath, 'effects'],
                index,
                numberOfSequencesWithThisNodePath: 0,
            },
            label: 'Effects',
            children: sequence.effects.map((effect, i) => {
                const effectNodePath = [...nodePath, 'effects', i];
                return {
                    kind: 'group',
                    nodePathInfo: {
                        nodePath: effectNodePath,
                        index,
                        numberOfSequencesWithThisNodePath: 0,
                    },
                    label: effect.definition.label,
                    children: (0, exports.getEffectSchemaLabels)(effect).map((label) => {
                        var _a;
                        return ({
                            kind: 'field',
                            nodePathInfo: {
                                nodePath: [...effectNodePath, label.key],
                                index,
                                numberOfSequencesWithThisNodePath: 0,
                            },
                            label: (_a = label.description) !== null && _a !== void 0 ? _a : label.key,
                            field: null,
                        });
                    }),
                };
            }),
        });
    }
    const controlFields = (0, studio_shared_1.getFieldsToShow)({
        schema: sequence.controls.schema,
        currentRuntimeValueDotNotation: sequence.controls.currentRuntimeValueDotNotation,
        getDragOverrides,
        getCodeValues,
        nodePath,
    });
    if (controlFields && controlFields.length > 0) {
        for (const f of controlFields) {
            roots.push({
                kind: 'field',
                nodePathInfo: {
                    nodePath: [...nodePath, 'controls', f.key],
                    index,
                    numberOfSequencesWithThisNodePath: 0,
                },
                label: (_a = f.description) !== null && _a !== void 0 ? _a : f.key,
                field: f,
            });
        }
    }
    return roots;
};
exports.buildTimelineTree = buildTimelineTree;
const flattenVisibleTreeNodes = ({ nodes, getIsExpanded, depth = 0, }) => {
    const out = [];
    for (const node of nodes) {
        out.push({ node, depth });
        if (node.kind === 'group' && getIsExpanded(node.nodePathInfo)) {
            out.push(...(0, exports.flattenVisibleTreeNodes)({
                nodes: node.children,
                getIsExpanded,
                depth: depth + 1,
            }));
        }
    }
    return out;
};
exports.flattenVisibleTreeNodes = flattenVisibleTreeNodes;
const getTreeRowHeight = (node) => {
    if (node.kind === 'field' && node.field) {
        return node.field.rowHeight;
    }
    return exports.TREE_GROUP_ROW_HEIGHT;
};
exports.getTreeRowHeight = getTreeRowHeight;
const getExpandedTrackHeight = ({ sequence, nodePathInfo, getIsExpanded, getCodeValues, }) => {
    const tree = (0, exports.buildTimelineTree)({
        sequence,
        nodePathInfo,
        // We assume that no drag overrides can change the timeline layout
        getDragOverrides: () => ({}),
        getCodeValues,
    });
    const flat = (0, exports.flattenVisibleTreeNodes)({ nodes: tree, getIsExpanded });
    if (flat.length === 0) {
        return exports.TIMELINE_TRACK_EXPANDED_HEIGHT;
    }
    const totalRowsHeight = flat.reduce((sum, { node }) => sum + (0, exports.getTreeRowHeight)(node), 0);
    const separators = Math.max(0, flat.length - 1);
    return totalRowsHeight + separators;
};
exports.getExpandedTrackHeight = getExpandedTrackHeight;
exports.TIMELINE_LAYER_HEIGHT_VIDEO = 75;
exports.TIMELINE_LAYER_HEIGHT_IMAGE = 50;
exports.TIMELINE_LAYER_HEIGHT_AUDIO = 25;
const getTimelineLayerHeight = (type) => {
    if (type === 'video') {
        return exports.TIMELINE_LAYER_HEIGHT_VIDEO;
    }
    if (type === 'image') {
        return exports.TIMELINE_LAYER_HEIGHT_IMAGE;
    }
    return exports.TIMELINE_LAYER_HEIGHT_AUDIO;
};
exports.getTimelineLayerHeight = getTimelineLayerHeight;
