// @remotion/studio/dist/components/Timeline/TimelineTracks.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineTracks = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const is_collapsed_1 = require("./is-collapsed");
const MaxTimelineTracks_1 = require("./MaxTimelineTracks");
const TimelineSequence_1 = require("./TimelineSequence");
const TimelineTimeIndicators_1 = require("./TimelineTimeIndicators");
const content = {
    paddingLeft: timeline_layout_1.TIMELINE_PADDING,
    paddingRight: timeline_layout_1.TIMELINE_PADDING,
    paddingTop: 1,
};
const timelineContent = {
    minHeight: '100%',
};
const getExpandedPlaceholderStyle = ({ sequence, nodePathInfo, getIsExpanded, getCodeValues, }) => ({
    height: (0, timeline_layout_1.getExpandedTrackHeight)({
        sequence,
        nodePathInfo,
        getIsExpanded,
        getCodeValues,
    }) + timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM,
});
const TimelineTracksInner = ({ timeline, hasBeenCut }) => {
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const { getCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const previewServerConnected = previewServerState.type === 'connected';
    const timelineStyle = (0, react_1.useMemo)(() => {
        return {
            ...timelineContent,
            width: 100 + '%',
        };
    }, []);
    return (jsx_runtime_1.jsxs("div", { style: timelineStyle, children: [
            jsx_runtime_1.jsxs("div", { style: content, children: [
                    jsx_runtime_1.jsx(TimelineTimeIndicators_1.TimelineTimePadding, {}), timeline.map((track) => {
                        if ((0, is_collapsed_1.isTrackHidden)(track)) {
                            return null;
                        }
                        const isExpanded = track.nodePathInfo !== null && getIsExpanded(track.nodePathInfo);
                        return (jsx_runtime_1.jsxs("div", { children: [
                                jsx_runtime_1.jsx("div", { style: {
                                        height: (0, timeline_layout_1.getTimelineLayerHeight)(track.sequence.type),
                                        marginBottom: timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM,
                                    }, children: jsx_runtime_1.jsx(TimelineSequence_1.TimelineSequence, { s: track.sequence }) }), isExpanded && track.nodePathInfo && previewServerConnected ? (jsx_runtime_1.jsx("div", { style: getExpandedPlaceholderStyle({
                                        sequence: track.sequence,
                                        nodePathInfo: track.nodePathInfo,
                                        getIsExpanded,
                                        getCodeValues,
                                    }) })) : null] }, track.sequence.id));
                    })] }), hasBeenCut ? jsx_runtime_1.jsx(MaxTimelineTracks_1.MaxTimelineTracksReached, {}) : null] }));
};
exports.TimelineTracks = react_1.default.memo(TimelineTracksInner);
