// @remotion/studio/dist/components/Timeline/TimelineFieldRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineFieldRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const call_api_1 = require("../call-api");
const Padder_1 = require("./Padder");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const fieldRowBase = {
    display: 'flex',
    alignItems: 'center',
    gap: 8,
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const fieldName = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const fieldLabelRow = {
    flex: '0 0 50%',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: 6,
};
const TimelineFieldRow = ({ field, validatedLocation, paddingLeft, nestedDepth, nodePath, schema, }) => {
    var _a, _b;
    const { getCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const { getDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { setDragOverrides, clearDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const codeValuesForOverride = getCodeValues(nodePath);
    const codeValue = (_a = codeValuesForOverride === null || codeValuesForOverride === void 0 ? void 0 : codeValuesForOverride[field.key]) !== null && _a !== void 0 ? _a : null;
    if (codeValue === null) {
        throw new Error('Unexpectedly got null code value for field' + field.key);
    }
    const dragOverrideValue = (0, react_1.useMemo)(() => {
        var _a;
        return nodePath === null
            ? undefined
            : ((_a = getDragOverrides(nodePath)) !== null && _a !== void 0 ? _a : {})[field.key];
    }, [getDragOverrides, nodePath, field.key]);
    const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
        codeValue,
        runtimeValue: field.currentRuntimeValue,
        dragOverrideValue,
        defaultValue: field.fieldSchema.default,
        shouldResortToDefaultValueIfUndefined: true,
    });
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const onSave = (0, react_1.useCallback)((value) => {
        if (!codeValuesForOverride || !validatedLocation || !nodePath) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!codeValue || !codeValue.canUpdate) {
            return Promise.reject(new Error('Cannot save'));
        }
        const defaultValue = field.fieldSchema.default !== undefined
            ? JSON.stringify(field.fieldSchema.default)
            : null;
        const stringifiedValue = JSON.stringify(value);
        if (value === codeValue.codeValue) {
            return Promise.resolve();
        }
        if (defaultValue === stringifiedValue &&
            codeValue.codeValue === undefined) {
            return Promise.resolve();
        }
        let previousUpdate;
        // Optimistic update to prevent flicker
        setCodeValues(nodePath, (prev) => {
            previousUpdate = prev;
            return (0, studio_shared_1.optimisticUpdateForCodeValues)({
                previous: prev,
                fieldKey: field.key,
                value,
                schema,
            });
        });
        return (0, call_api_1.callApi)('/api/save-sequence-props', {
            fileName: validatedLocation.source,
            nodePath,
            key: field.key,
            value: stringifiedValue,
            defaultValue,
            schema,
        })
            .then((data) => {
            setCodeValues(nodePath, () => data);
        })
            .catch(() => {
            // In case something went wrong, undo optimistic update
            if (previousUpdate) {
                setCodeValues(nodePath, (current) => {
                    if (previousUpdate) {
                        return previousUpdate;
                    }
                    return current;
                });
            }
        });
    }, [
        codeValue,
        codeValuesForOverride,
        field.fieldSchema.default,
        field.key,
        nodePath,
        schema,
        setCodeValues,
        validatedLocation,
    ]);
    const onDragValueChange = (0, react_1.useCallback)((value) => {
        if (nodePath === null) {
            throw new Error('Cannot drag value');
        }
        setDragOverrides(nodePath, field.key, value);
    }, [setDragOverrides, nodePath, field.key]);
    const onDragEnd = (0, react_1.useCallback)(() => {
        if (nodePath === null) {
            throw new Error('Cannot clear drag value');
        }
        clearDragOverrides(nodePath);
    }, [clearDragOverrides, nodePath]);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...fieldRowBase,
            height: field.rowHeight,
            paddingLeft,
        };
    }, [field.rowHeight, paddingLeft]);
    return (jsx_runtime_1.jsxs("div", { style: style, children: [
            jsx_runtime_1.jsx(Padder_1.Padder, { depth: nestedDepth + 1 }), jsx_runtime_1.jsx("div", { style: fieldLabelRow, children: jsx_runtime_1.jsx("span", { style: fieldName, children: (_b = field.description) !== null && _b !== void 0 ? _b : field.key }) }), jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineFieldValue, { field: field, propStatus: codeValue, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, effectiveValue: effectiveValue })
        ] }));
};
exports.TimelineFieldRow = TimelineFieldRow;
