// @remotion/studio/dist/components/Timeline/TimelineVideoInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineVideoInfo = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const timeline_utils_1 = require("@remotion/timeline-utils");
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const AudioWaveform_1 = require("../AudioWaveform");
const FILMSTRIP_HEIGHT = timeline_layout_1.TIMELINE_LAYER_HEIGHT_IMAGE - 2;
const outerStyle = {
    width: '100%',
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
};
const filmstripContainerStyle = {
    height: FILMSTRIP_HEIGHT,
    width: '100%',
    backgroundColor: 'rgba(0, 0, 0, 0.3)',
    display: 'flex',
    borderTopLeftRadius: 2,
    fontSize: 10,
    fontFamily: 'Arial, Helvetica',
};
const TimelineVideoInfo = ({ src, visualizationWidth, naturalWidth, trimBefore, durationInFrames, playbackRate, volume, doesVolumeChange, premountWidth, postmountWidth, loopDisplay, }) => {
    const { fps } = (0, remotion_1.useVideoConfig)();
    const ref = (0, react_1.useRef)(null);
    const [error, setError] = (0, react_1.useState)(null);
    const aspectRatio = (0, react_1.useRef)((0, timeline_utils_1.getAspectRatioFromCache)(src));
    // for rendering frames
    (0, react_1.useEffect)(() => {
        if (error) {
            return;
        }
        const { current } = ref;
        if (!current) {
            return;
        }
        const controller = new AbortController();
        const canvas = document.createElement('canvas');
        canvas.width = visualizationWidth;
        canvas.height = FILMSTRIP_HEIGHT;
        const ctx = canvas.getContext('2d');
        if (!ctx) {
            return;
        }
        current.appendChild(canvas);
        const loopWidth = (0, timeline_utils_1.getLoopDisplayWidth)({
            visualizationWidth: naturalWidth,
            loopDisplay,
        });
        const shouldRepeatVideo = (0, timeline_utils_1.shouldTileLoopDisplay)(loopDisplay);
        const targetCanvas = shouldRepeatVideo
            ? document.createElement('canvas')
            : canvas;
        targetCanvas.width = shouldRepeatVideo
            ? Math.max(1, Math.ceil(loopWidth))
            : canvas.width;
        targetCanvas.height = canvas.height;
        const targetCtx = shouldRepeatVideo ? targetCanvas.getContext('2d') : ctx;
        if (!targetCtx) {
            current.removeChild(canvas);
            return;
        }
        const repeatTarget = () => {
            if (!shouldRepeatVideo) {
                return;
            }
            const pattern = ctx.createPattern(targetCanvas, 'repeat-x');
            if (!pattern) {
                return;
            }
            pattern.setTransform(new DOMMatrix().scaleSelf(loopWidth / targetCanvas.width, 1));
            ctx.clearRect(0, 0, canvas.width, canvas.height);
            ctx.fillStyle = pattern;
            ctx.fillRect(0, 0, canvas.width, canvas.height);
        };
        // desired-timestamp -> filled-timestamp
        const filledSlots = new Map();
        const fromSeconds = trimBefore / fps;
        const visibleDurationInFrames = shouldRepeatVideo && loopDisplay
            ? loopDisplay.durationInFrames
            : durationInFrames;
        // Trim is applied first, then playbackRate. Each composition frame
        // advances the source video by `playbackRate` source frames.
        const toSeconds = fromSeconds + (visibleDurationInFrames * playbackRate) / fps;
        const targetWidth = shouldRepeatVideo ? targetCanvas.width : naturalWidth;
        if (aspectRatio.current !== null) {
            (0, timeline_utils_1.ensureSlots)({
                filledSlots,
                naturalWidth: targetWidth,
                fromSeconds,
                toSeconds,
                aspectRatio: aspectRatio.current,
                frameHeight: FILMSTRIP_HEIGHT,
            });
            (0, timeline_utils_1.fillWithCachedFrames)({
                ctx: targetCtx,
                naturalWidth: targetWidth,
                filledSlots,
                src,
                segmentDuration: toSeconds - fromSeconds,
                fromSeconds,
                devicePixelRatio: window.devicePixelRatio,
                frameHeight: FILMSTRIP_HEIGHT,
            });
            repeatTarget();
            const unfilled = Array.from(filledSlots.keys()).filter((timestamp) => !filledSlots.get(timestamp));
            // Don't extract frames if all slots are filled
            if (unfilled.length === 0) {
                return () => {
                    current.removeChild(canvas);
                };
            }
        }
        (0, timeline_utils_1.extractFrames)({
            timestampsInSeconds: ({ track, }) => {
                aspectRatio.current = track.width / track.height;
                timeline_utils_1.aspectRatioCache.set(src, aspectRatio.current);
                (0, timeline_utils_1.ensureSlots)({
                    filledSlots,
                    fromSeconds,
                    toSeconds,
                    naturalWidth: targetWidth,
                    aspectRatio: aspectRatio.current,
                    frameHeight: FILMSTRIP_HEIGHT,
                });
                return Array.from(filledSlots.keys()).map((timestamp) => timestamp / timeline_utils_1.WEBCODECS_TIMESCALE);
            },
            src,
            onVideoSample: (sample) => {
                let frame;
                try {
                    frame = sample.toVideoFrame();
                    const scale = (FILMSTRIP_HEIGHT / frame.displayHeight) * window.devicePixelRatio;
                    const transformed = (0, timeline_utils_1.resizeVideoFrame)({
                        frame,
                        scale,
                    });
                    if (transformed !== frame) {
                        frame.close();
                    }
                    frame = undefined;
                    const databaseKey = (0, timeline_utils_1.makeFrameDatabaseKey)(src, transformed.timestamp);
                    (0, timeline_utils_1.addFrameToCache)(databaseKey, transformed);
                    if (aspectRatio.current === null) {
                        throw new Error('Aspect ratio is not set');
                    }
                    (0, timeline_utils_1.ensureSlots)({
                        filledSlots,
                        fromSeconds,
                        toSeconds,
                        naturalWidth: targetWidth,
                        aspectRatio: aspectRatio.current,
                        frameHeight: FILMSTRIP_HEIGHT,
                    });
                    (0, timeline_utils_1.fillFrameWhereItFits)({
                        ctx: targetCtx,
                        filledSlots,
                        visualizationWidth: targetWidth,
                        frame: transformed,
                        segmentDuration: toSeconds - fromSeconds,
                        fromSeconds,
                        devicePixelRatio: window.devicePixelRatio,
                        frameHeight: FILMSTRIP_HEIGHT,
                    });
                    repeatTarget();
                }
                catch (e) {
                    if (frame) {
                        frame.close();
                    }
                    throw e;
                }
                finally {
                    sample.close();
                }
            },
            signal: controller.signal,
        })
            .then(() => {
            if (controller.signal.aborted) {
                return;
            }
            (0, timeline_utils_1.fillWithCachedFrames)({
                ctx: targetCtx,
                naturalWidth: targetWidth,
                filledSlots,
                src,
                segmentDuration: toSeconds - fromSeconds,
                fromSeconds,
                devicePixelRatio: window.devicePixelRatio,
                frameHeight: FILMSTRIP_HEIGHT,
            });
            repeatTarget();
        })
            .catch((e) => {
            setError(e);
        });
        return () => {
            controller.abort();
            current.removeChild(canvas);
        };
    }, [
        durationInFrames,
        error,
        fps,
        loopDisplay,
        naturalWidth,
        playbackRate,
        src,
        trimBefore,
        visualizationWidth,
    ]);
    const audioWidth = visualizationWidth - premountWidth - postmountWidth;
    const audioStyle = (0, react_1.useMemo)(() => {
        return {
            height: timeline_layout_1.TIMELINE_LAYER_HEIGHT_AUDIO,
            width: audioWidth,
            position: 'relative',
            marginLeft: premountWidth,
        };
    }, [audioWidth, premountWidth]);
    return (jsx_runtime_1.jsxs("div", { style: outerStyle, children: [
            jsx_runtime_1.jsx("div", { ref: ref, style: filmstripContainerStyle }), jsx_runtime_1.jsx("div", { style: audioStyle, children: jsx_runtime_1.jsx(AudioWaveform_1.AudioWaveform, { src: src, visualizationWidth: audioWidth, startFrom: trimBefore, durationInFrames: durationInFrames, volume: volume, doesVolumeChange: doesVolumeChange, playbackRate: playbackRate, loopDisplay: loopDisplay }) })
        ] }));
};
exports.TimelineVideoInfo = TimelineVideoInfo;
