// @remotion/studio/dist/components/Timeline/use-sequence-props-subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSequencePropsSubscription = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const sequence_props_subscription_store_1 = require("./sequence-props-subscription-store");
const useSequencePropsSubscription = ({ originalLocation, overrideId, schema, }) => {
    var _a, _b, _c;
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { setOverrideIdToNodePath } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsSettersContext);
    const { previewServerState: state } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const clientId = state.type === 'connected' ? state.clientId : undefined;
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const locationSource = (_a = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source) !== null && _a !== void 0 ? _a : null;
    const locationLine = (_b = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.line) !== null && _b !== void 0 ? _b : null;
    const locationColumn = (_c = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.column) !== null && _c !== void 0 ? _c : null;
    (0, react_1.useEffect)(() => {
        if (!clientId ||
            !locationSource ||
            !locationLine ||
            locationColumn === null ||
            !schema) {
            return;
        }
        const { release } = (0, sequence_props_subscription_store_1.acquireSequencePropsSubscription)({
            fileName: locationSource,
            line: locationLine,
            column: locationColumn,
            schema,
            clientId,
            applyOnce: (result) => {
                if (!result.success) {
                    return;
                }
                setCodeValues(result.nodePath, () => result.status);
            },
            applyEach: (result) => {
                if (!result.success) {
                    return;
                }
                setOverrideIdToNodePath(overrideId, result.nodePath);
            },
        });
        return () => {
            release();
        };
    }, [
        clientId,
        locationColumn,
        locationLine,
        locationSource,
        overrideId,
        schema,
        setCodeValues,
        setOverrideIdToNodePath,
    ]);
};
exports.useSequencePropsSubscription = useSequencePropsSubscription;
