// @remotion/transitions/dist/html-in-canvas-presentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeHtmlInCanvasPresentation = exports.HtmlInCanvasPresentation = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const remotion_2 = require("remotion");
const HtmlInCanvasPresentation = ({ children, onElementImage, onUnmount, presentationProgress, presentationDirection, shader, _experimentalEffects, passedProps, bothEnteringAndExiting, }) => {
    if (!remotion_1.HtmlInCanvas.isSupported()) {
        throw new Error('HTML in Canvas is not supported. Open this page in Chrome Canary with chrome://flags/#canvas-draw-element enabled.');
    }
    const canvasRef = (0, react_1.useRef)(null);
    const canvasSubtreeStyle = (0, react_1.useMemo)(() => {
        return {
            width: '100%',
            height: '100%',
            position: 'absolute',
            top: 0,
            left: 0,
            right: 0,
            bottom: 0,
        };
    }, []);
    const [offscreenCanvas] = (0, react_1.useState)(() => new OffscreenCanvas(1, 1));
    const passedPropsRef = (0, react_1.useRef)(passedProps);
    passedPropsRef.current = passedProps;
    const frame = (0, remotion_2.useCurrentFrame)();
    const frameRef = (0, react_1.useRef)(frame);
    frameRef.current = frame;
    const memoizedEffects = remotion_2.Internals.useMemoizedEffects(remotion_2.Internals.flattenEffects(_experimentalEffects !== null && _experimentalEffects !== void 0 ? _experimentalEffects : []));
    const effectsRef = (0, react_1.useRef)(memoizedEffects);
    effectsRef.current = memoizedEffects;
    const [instance] = (0, react_1.useState)(() => shader(offscreenCanvas));
    (0, react_1.useLayoutEffect)(() => {
        return () => {
            instance.cleanup();
        };
    }, [offscreenCanvas, instance]);
    const chainState = remotion_2.Internals.useEffectChainState();
    const { delayRender, continueRender } = (0, remotion_1.useDelayRender)();
    const draw = (0, react_1.useCallback)(async (prevImage, nextImage, progress) => {
        var _a, _b, _c, _d, _e;
        if (!canvasRef.current) {
            throw new Error('Canvas not found');
        }
        const handle = delayRender('onPaint');
        if (!prevImage && !nextImage) {
            continueRender(handle);
            instance.clear();
            return;
        }
        const width = (_b = (_a = prevImage === null || prevImage === void 0 ? void 0 : prevImage.width) !== null && _a !== void 0 ? _a : nextImage === null || nextImage === void 0 ? void 0 : nextImage.width) !== null && _b !== void 0 ? _b : 0;
        const height = (_d = (_c = prevImage === null || prevImage === void 0 ? void 0 : prevImage.height) !== null && _c !== void 0 ? _c : nextImage === null || nextImage === void 0 ? void 0 : nextImage.height) !== null && _d !== void 0 ? _d : 0;
        if (width === 0 || height === 0) {
            continueRender(handle);
            instance.clear();
            return;
        }
        offscreenCanvas.width = width;
        offscreenCanvas.height = height;
        instance.draw({
            prevImage,
            nextImage,
            width,
            height,
            time: progress,
            passedProps: passedPropsRef.current,
        });
        await remotion_2.Internals.runEffectChain({
            state: chainState.get(width, height),
            source: offscreenCanvas,
            effects: (_e = effectsRef.current) !== null && _e !== void 0 ? _e : [],
            frame: frameRef.current,
            width,
            height,
            output: canvasRef.current,
        });
        continueRender(handle);
    }, [chainState, instance, offscreenCanvas, continueRender, delayRender]);
    const passThrough = bothEnteringAndExiting && presentationDirection === 'exiting';
    (0, react_1.useLayoutEffect)(() => {
        if (passThrough) {
            return;
        }
        const canvas = canvasRef.current;
        if (!canvas) {
            throw new Error('Canvas not found');
        }
        canvas.layoutSubtree = true;
        const onPaint = () => {
            const firstChild = canvas.firstChild;
            if (!firstChild) {
                return;
            }
            const elementImage = canvas.captureElementImage(firstChild);
            onElementImage(elementImage, draw);
        };
        canvas.addEventListener('paint', onPaint);
        return () => {
            canvas.removeEventListener('paint', onPaint);
        };
    }, [onElementImage, presentationDirection, draw, passThrough]);
    (0, react_1.useLayoutEffect)(() => {
        var _a;
        if (passThrough) {
            return;
        }
        const canvas = canvasRef.current;
        if (!canvas) {
            throw new Error('Canvas not found');
        }
        (_a = canvas.requestPaint) === null || _a === void 0 ? void 0 : _a.call(canvas);
    }, [presentationProgress, passThrough, memoizedEffects]);
    (0, react_1.useLayoutEffect)(() => {
        if (passThrough) {
            return;
        }
        return () => {
            onUnmount();
        };
    }, [onUnmount, passThrough]);
    (0, react_1.useLayoutEffect)(() => {
        if (passThrough) {
            return;
        }
        const canvas = canvasRef.current;
        if (!canvas) {
            return;
        }
        // Size the canvas grid to match the device scale factor to prevent blurriness.
        const observer = new ResizeObserver(([entry]) => {
            canvas.width = entry.devicePixelContentBoxSize[0].inlineSize;
            canvas.height = entry.devicePixelContentBoxSize[0].blockSize;
        });
        observer.observe(canvas, { box: 'device-pixel-content-box' });
    }, [passThrough]);
    if (passThrough) {
        return children;
    }
    return (jsx_runtime_1.jsx(remotion_2.AbsoluteFill, { children: jsx_runtime_1.jsx("canvas", { ref: canvasRef, style: canvasSubtreeStyle, children: children }) }));
};
exports.HtmlInCanvasPresentation = HtmlInCanvasPresentation;
const makeHtmlInCanvasPresentation = (shader) => {
    const CompWithShader = (props) => {
        const { passedProps, ...otherProps } = props;
        const { _experimentalEffects, ...restPassedProps } = props.passedProps;
        return (jsx_runtime_1.jsx(exports.HtmlInCanvasPresentation, { shader: shader, passedProps: restPassedProps, _experimentalEffects: _experimentalEffects, ...otherProps }));
    };
    return (props) => {
        return {
            component: CompWithShader,
            props,
        };
    };
};
exports.makeHtmlInCanvasPresentation = makeHtmlInCanvasPresentation;
