// @remotion/studio-server/dist/preview-server/routes/log-updates/formatting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatAddition = exports.formatDeletion = exports.formatPropDelta = exports.addedPrefixIfNoColor = exports.strikeThroughOrRemovedPrefix = exports.strikeThrough = exports.numberValue = exports.stringValue = exports.punctuation = exports.equals = exports.attrName = exports.bg = exports.fg = exports.colorValue = exports.colorEnabled = void 0;
const renderer_1 = require("@remotion/renderer");
const colorEnabled = () => renderer_1.RenderInternals.chalk.enabled();
exports.colorEnabled = colorEnabled;
const colorValue = (str) => {
    if ((str.startsWith("'") && str.endsWith("'")) ||
        (str.startsWith('"') && str.endsWith('"'))) {
        return (0, exports.stringValue)(str);
    }
    if (/^-?\d+(\.\d+)?$/.test(str)) {
        return (0, exports.numberValue)(str);
    }
    return (0, exports.punctuation)(str);
};
exports.colorValue = colorValue;
// eslint-disable-next-line no-control-regex
const stripAnsi = (str) => str.replace(/\u001b\[[0-9;]*m/g, '');
// 24-bit ANSI helpers
const fg = (r, g, b, str) => (0, exports.colorEnabled)() ? `\u001b[38;2;${r};${g};${b}m${str}\u001b[39m` : str;
exports.fg = fg;
const bg = (r, g, b, str) => (0, exports.colorEnabled)() ? `\u001b[48;2;${r};${g};${b}m${str}\u001b[49m` : str;
exports.bg = bg;
// Monokai-inspired syntax colors
const attrName = (str) => (0, exports.fg)(166, 226, 46, str);
exports.attrName = attrName;
const equals = (str) => (0, exports.fg)(249, 38, 114, str);
exports.equals = equals;
const punctuation = (str) => (0, exports.fg)(248, 248, 242, str);
exports.punctuation = punctuation;
const stringValue = (str) => (0, exports.fg)(230, 219, 116, str);
exports.stringValue = stringValue;
const numberValue = (str) => (0, exports.fg)(174, 129, 255, str);
exports.numberValue = numberValue;
const strikeThrough = (str) => `\u001b[9m\u001b[38;2;255;85;85m${stripAnsi(str)}\u001b[39m\u001b[29m`;
exports.strikeThrough = strikeThrough;
const strikeThroughOrRemovedPrefix = (str) => (0, exports.colorEnabled)() ? (0, exports.strikeThrough)(str) : 'removed: ' + str;
exports.strikeThroughOrRemovedPrefix = strikeThroughOrRemovedPrefix;
const addedPrefixIfNoColor = (str) => (0, exports.colorEnabled)() ? str : 'added: ' + stripAnsi(str);
exports.addedPrefixIfNoColor = addedPrefixIfNoColor;
// Format key={value} with Monokai syntax highlighting
const formatSimpleProp = (key, value) => {
    return `${(0, exports.attrName)(key)}${(0, exports.equals)('=')}${(0, exports.punctuation)('{')}${(0, exports.colorValue)(value)}${(0, exports.punctuation)('}')}`;
};
// Format parentKey={{childKey: value}} with Monokai syntax highlighting
const formatNestedProp = (parentKey, childKey, value) => {
    return `${(0, exports.attrName)(parentKey)}${(0, exports.equals)('=')}${(0, exports.punctuation)('{{')}${(0, exports.punctuation)(childKey)}${(0, exports.punctuation)(':')} ${(0, exports.colorValue)(value)}${(0, exports.punctuation)('}}')}`;
};
const formatPropDelta = ({ key, valueString }) => {
    const dotIdx = key.indexOf('.');
    if (dotIdx === -1) {
        return formatSimpleProp(key, valueString);
    }
    return formatNestedProp(key.slice(0, dotIdx), key.slice(dotIdx + 1), valueString);
};
exports.formatPropDelta = formatPropDelta;
const formatDeletion = (prop) => {
    const formatted = (0, exports.formatPropDelta)(prop);
    return (0, exports.strikeThroughOrRemovedPrefix)(formatted);
};
exports.formatDeletion = formatDeletion;
const formatAddition = (prop) => {
    const formatted = (0, exports.formatPropDelta)(prop);
    return (0, exports.addedPrefixIfNoColor)(formatted);
};
exports.formatAddition = formatAddition;
