// @remotion/studio-server/dist/preview-server/sequence-props-watchers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unsubscribeClientSequencePropsWatchers = exports.unsubscribeFromSequencePropsWatchers = exports.subscribeToSequencePropsWatchers = void 0;
const node_path_1 = __importDefault(require("node:path"));
const file_watcher_1 = require("../file-watcher");
const live_events_1 = require("./live-events");
const node_path_cache_1 = require("./node-path-cache");
const can_update_sequence_props_1 = require("./routes/can-update-sequence-props");
const sequencePropsWatchers = {};
const makeWatcherKey = ({ absolutePath, nodePath, }) => {
    return `${absolutePath}:${JSON.stringify(nodePath)}`;
};
const getSequencePropsStatus = ({ fileName, line, column, keys, remotionRoot, }) => {
    // Try cached nodePath first (handles stale source maps after suppressed rebuilds)
    const cachedNodePath = (0, node_path_cache_1.getCachedNodePath)(fileName, line, column);
    if (cachedNodePath) {
        const cachedResult = (0, can_update_sequence_props_1.computeSequencePropsStatus)({
            fileName,
            nodePath: cachedNodePath,
            keys,
            remotionRoot,
        });
        if (cachedResult.canUpdate) {
            return { status: cachedResult, nodePath: cachedNodePath, success: true };
        }
    }
    const status = (0, can_update_sequence_props_1.computeSequencePropsStatusFromFilenameByLine)({
        fileName,
        line,
        keys,
        remotionRoot,
    });
    return status;
};
const subscribeToSequencePropsWatchers = ({ fileName, line, column, keys, remotionRoot, clientId, }) => {
    var _a;
    const initialResult = getSequencePropsStatus({
        fileName,
        line,
        column,
        keys,
        remotionRoot,
    });
    if (!initialResult.success) {
        return initialResult;
    }
    const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
    // Cache the resolved nodePath for future lookups with stale source maps
    (0, node_path_cache_1.setCachedNodePath)(fileName, line, column, initialResult.nodePath);
    const { nodePath } = initialResult;
    const watcherKey = makeWatcherKey({ absolutePath, nodePath });
    // If a watcher already exists for this key, just bump the ref count
    if ((_a = sequencePropsWatchers[clientId]) === null || _a === void 0 ? void 0 : _a[watcherKey]) {
        sequencePropsWatchers[clientId][watcherKey].refCount++;
        return initialResult;
    }
    const { unwatch } = (0, file_watcher_1.installFileWatcher)({
        file: absolutePath,
        existenceOnly: false,
        onChange: (event) => {
            if (event.type === 'deleted') {
                return;
            }
            let result;
            try {
                result = (0, can_update_sequence_props_1.computeSequencePropsStatusFromContent)(event.content, nodePath, keys);
            }
            catch (_a) {
                return;
            }
            (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
                listener.sendEventToClientId(clientId, {
                    type: 'sequence-props-updated',
                    fileName,
                    nodePath,
                    result,
                });
            });
        },
    });
    if (!sequencePropsWatchers[clientId]) {
        sequencePropsWatchers[clientId] = {};
    }
    sequencePropsWatchers[clientId][watcherKey] = { unwatch, refCount: 1 };
    return initialResult;
};
exports.subscribeToSequencePropsWatchers = subscribeToSequencePropsWatchers;
const unsubscribeFromSequencePropsWatchers = ({ fileName, nodePath, remotionRoot, clientId, }) => {
    const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
    const watcherKey = makeWatcherKey({ absolutePath, nodePath });
    if (!sequencePropsWatchers[clientId] ||
        !sequencePropsWatchers[clientId][watcherKey]) {
        return;
    }
    sequencePropsWatchers[clientId][watcherKey].refCount--;
    if (sequencePropsWatchers[clientId][watcherKey].refCount <= 0) {
        sequencePropsWatchers[clientId][watcherKey].unwatch();
        delete sequencePropsWatchers[clientId][watcherKey];
    }
};
exports.unsubscribeFromSequencePropsWatchers = unsubscribeFromSequencePropsWatchers;
const unsubscribeClientSequencePropsWatchers = (clientId) => {
    if (!sequencePropsWatchers[clientId]) {
        return;
    }
    Object.values(sequencePropsWatchers[clientId]).forEach((watcher) => {
        watcher.unwatch();
    });
    delete sequencePropsWatchers[clientId];
};
exports.unsubscribeClientSequencePropsWatchers = unsubscribeClientSequencePropsWatchers;
